### PR TITLE
Add more annotations to `java.lang.ref`.

### PR DESCRIPTION
**Pull request by cpovirk**

### CODE REVIEW

File: src/java.base/share/classes/java/lang/ref/Reference.java [hunk 1]
```diff
@@ -520,7 +520,7 @@ protected Object clone() throws CloneNotSupportedException {
      * @since 9
      */
     @ForceInline
-    public static void reachabilityFence(Object ref) {
+    public static void reachabilityFence(@Nullable Object ref) {
```
Comment (wmdietl) [inline src/java.base/share/classes/java/lang/ref/Reference.java:523]: What use is there to allow null here? I agree that the documentation says that null is allowed, so no NPE will be thrown. On the other hand, a reachability fence only makes sense for a real object. Doesn't allowing null here simply allow likely buggy code?
Can you add a `CFComment` that adds a rationale for making this parameter nullable?

**Review by wmdietl — COMMENTED** [review]
Thanks! The three changes to constructors are all consistent with the javadoc and make sense to me. I have a question about the `reachabilityFence` change.

File: src/java.base/share/classes/java/lang/ref/Reference.java (hunk 1 above)
Comment (cpovirk) [reply src/java.base/share/classes/java/lang/ref/Reference.java]: Only the constructor changes were motivated by real code; I just noticed `reachabilityFence` while I was in the area.

This method is rarely called. And, at least inside Google, it's always called with an argument of `this`. So the choice of annotation is probably irrelevant in practice. I've removed `@Nullable` and added a `@CFComment` about this.

**Review by wmdietl — APPROVED** [review]
Thanks!

**Pull request merged**